serializers/shared/src/main/java/io/pravega/schemaregistry/serializer/shared/impl/AbstractDeserializer.java

```diff
@@ -93,8 +93,17 @@ public T deserialize(ByteBuffer data) {
                 decoded = decoders.decode(encodingInfo.getCodecType(), data);
             }
 
-            inputStream = new ByteArrayInputStream(decoded.array(), 
-                    decoded.arrayOffset() + decoded.position(), decoded.remaining());
+            byte[] b;
+            if (decoded.hasArray()) {
+                b = decoded.array();
+                start = decoded.arrayOffset() + decoded.position();
+            } else {
+                b = new byte[decoded.remaining()];
+                decoded.get(b);
+                start = decoded.position();
+            }
+
+            inputStream = new ByteArrayInputStream(b, start, decoded.remaining());
             // pass writer schema for schema to be read into
             readerSchema = schemaInfo == null ? writerSchema : schemaInfo;
```

Comment:
nit: The condition can be simplified by doing bt collocating the variable used in the condition and the statement where it is used. `schemaInfo != null ? schemaInfo: writerSchema`